### PR TITLE
Enable phpstan in DBAL tests

### PR DESCRIPTION
**Pull request by jwage**
|      Q       |   A
|------------- | -----------
| Type         | improvement
| BC Break     | no

#### Summary

Enable phpstan in DBAL tests.

### CODE REVIEW

File: tests/Doctrine/Tests/DBAL/Sharding/PoolingShardConnectionTest.php [hunk 1]
```diff
@@ -239,9 +248,12 @@ public function testGetParamsOverride() : void
      */
     private function createConnection(array $parameters) : PoolingShardConnection
     {
-        return DriverManager::getConnection(array_merge(
+        /** @var PoolingShardConnection $connection */
```
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Sharding/PoolingShardConnectionTest.php]: Can it be a PHPUnit assertion?

File: tests/Doctrine/Tests/DBAL/Schema/SchemaTest.php [hunk 2]
```diff
@@ -462,6 +464,8 @@ public function testVisitsNamespaceVisitor() : void
         $visitor->expects($this->exactly(2))
             ->method('acceptSequence');
 
-        self::assertNull($schema->visit($visitor));
+        $schema->visit($visitor);
+
+        self::doesNotPerformAssertions(); // FIXME
```
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Schema/SchemaTest.php]: The visitor above performs some assertions. Why is this needed?

File: lib/Doctrine/DBAL/Query/Expression/ExpressionBuilder.php [hunk 3]
```diff
@@ -236,8 +236,8 @@ public function notLike(string $x, $y/*, ?string $escapeChar = null */) : string
     /**
      * Creates a IN () comparison expression with the given arguments.
      *
-     * @param string          $x The field in string format to be inspected by IN() comparison.
-     * @param string|string[] $y The placeholder or the array of values to be used by IN() comparison.
+     * @param string         $x The field in string format to be inspected by IN() comparison.
+     * @param string|mixed[] $y The placeholder or the array of values to be used by IN() comparison.
```
[the remark this answers is not in the document]
Comment (jwage) [reply lib/Doctrine/DBAL/Query/Expression/ExpressionBuilder.php]: Should I go ahead and change all the other `mixed` to `string` then too?

File: tests/Doctrine/Tests/DBAL/Driver/StatementIteratorTest.php [hunk 4]
```diff
@@ -43,7 +44,9 @@ public function testIteratorIterationCallsFetchOncePerStep() : void
 
         $stmtIterator = new StatementIterator($stmt);
 
-        $this->assertIterationCallsFetchOncePerStep($stmtIterator, $calls);
+        foreach ($stmtIterator as $i => $_) {
```
[the remark this answers is not in the document]
Comment (jwage) [reply tests/Doctrine/Tests/DBAL/Driver/StatementIteratorTest.php]: It was sharing a method where it passed multiple different types of objects through and I couldn't get the type to work for the mock and the concrete implementation.

File: tests/Doctrine/Tests/DBAL/Functional/DataAccessTest.php [hunk 5]
```diff
@@ -475,7 +477,7 @@ public function testPrepareQueryBindValueDateTimeType() : void
     {
         $sql  = 'SELECT count(*) AS c FROM fetch_table WHERE test_datetime = ?';
         $stmt = $this->connection->prepare($sql);
-        $stmt->bindValue(1, new DateTime('2010-01-01 10:10:10'), Types::DATETIME_MUTABLE);
+        $stmt->bindValue(1, '2010-01-01 10:10:10', ParameterType::STRING);
```
[the remark this answers is not in the document]
Comment (jwage) [reply tests/Doctrine/Tests/DBAL/Functional/DataAccessTest.php]: What type should it be? I got an error about not being able to cast DateTime to a string.

File: tests/Doctrine/Tests/DBAL/Functional/Driver/IBMDB2/DB2StatementTest.php [hunk 6]
```diff
@@ -29,6 +31,7 @@ protected function setUp() : void
     public function testExecutionErrorsAreNotSuppressed() : void
     {
         $stmt = $this->connection->prepare('SELECT * FROM SYSIBM.SYSDUMMY1 WHERE \'foo\' = ?');
+        assert($stmt instanceof Statement);
```
[the remark this answers is not in the document]
Comment (jwage) [reply tests/Doctrine/Tests/DBAL/Functional/Driver/IBMDB2/DB2StatementTest.php]: That link doesn't seem to deep link properly.

Clarify the return type in `Doctrine\DBAL\Connection::prepare()` ?

File: tests/Doctrine/Tests/DBAL/Functional/Schema/SQLServerSchemaManagerTest.php [hunk 7]
```diff
@@ -29,7 +29,7 @@ public function testDropColumnConstraints() : void
 
         $this->schemaManager->createTable($table);
 
-        $diff = new TableDiff('sqlsrv_drop_column', [], [], [new Column('todrop', Type::getType('decimal'))]);
+        $diff = new TableDiff('sqlsrv_drop_column', [], [], ['todrop' => new Column('todrop', Type::getType('decimal'))]);
```
[the remark this answers is not in the document]
Comment (jwage) [reply tests/Doctrine/Tests/DBAL/Functional/Schema/SQLServerSchemaManagerTest.php]: We can try it. The key of the array is used in some cases. For example, to check if a column exists by checking the key of the array.

```php
$table->addColumn('`comment_quoted`', 'integer', ['comment' => 'Doctrine 0wnz comments for explicitly quoted columns!']);
```

The column name is normalized and quotes/ticks are trimmed before putting it in the columns array. Then this usage exists:

```php
    /**
     * Drops a Column from the Table.
     */
    public function dropColumn(string $columnName) : self
    {
        $columnName = $this->normalizeIdentifier($columnName);

        unset($this->_columns[$columnName]);

        return $this;
    }
```

So it is possible probably, would just have to refactor...move the normalization to the `Column` API or have a `Columns` object or something where we implement finding columns, removing columns, etc.

File: tests/Doctrine/Tests/DBAL/Tools/DumperTest.php [hunk 8]
```diff
@@ -100,8 +100,8 @@ public function testExportParentAttributes(TestAsset\ParentClass $class, array $
         $print_r_class    = print_r($class, true);
         $print_r_expected = print_r($expected, true);
 
-        $print_r_class    = substr($print_r_class, strpos($print_r_class, '('));
-        $print_r_expected = substr($print_r_expected, strpos($print_r_expected, '('));
+        $print_r_class    = substr($print_r_class, (int) strpos($print_r_class, '('));
+        $print_r_expected = substr($print_r_expected, (int) strpos($print_r_expected, '('));
```
[the remark this answers is not in the document]
Comment (jwage) [reply tests/Doctrine/Tests/DBAL/Tools/DumperTest.php]: I really don't know either :(

[the remark this answers is not in the document]
Comment (Ocramius) [reply tests/Doctrine/Tests/DBAL/Tools/DumperTest.php]: Looks like a property-based test, kina

**Comment by Ocramius** [comment]
Failures are only in CS:

```
FILE: tests/Doctrine/Tests/DBAL/Types/JsonTest.php
--------------------------------------------------------------------------------
FOUND 2 ERRORS AFFECTING 2 LINES
--------------------------------------------------------------------------------
 15 | ERROR | [x] Type RuntimeException is not used in this file.
    |       |     (SlevomatCodingStandard.Namespaces.UnusedUses.UnusedUse)
 18 | ERROR | [x] Type json_encode is not used in this file.
    |       |     (SlevomatCodingStandard.Namespaces.UnusedUses.UnusedUse)
--------------------------------------------------------------------------------
PHPCBF CAN FIX THE 2 MARKED SNIFF VIOLATIONS AUTOMATICALLY
--------------------------------------------------------------------------------
FILE: tests/Doctrine/Tests/DBAL/Functional/TemporaryTableTest.php
--------------------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
--------------------------------------------------------------------------------
 15 | ERROR | Class TemporaryTableTest contains write-only property
    |       | $tempTableCreated.
    |       | (SlevomatCodingStandard.Classes.UnusedPrivateElements.WriteOnlyProperty)
--------------------------------------------------------------------------------
```

File: tests/Doctrine/Tests/DBAL/Functional/Schema/SchemaManagerFunctionalTestCase.php [hunk 9]
```diff
@@ -1431,6 +1447,8 @@ public function testPrimaryKeyAutoIncrement() : void
         $this->connection->insert('test_pk_auto_increment', ['text' => '1']);
 
         $query = $this->connection->query('SELECT id FROM test_pk_auto_increment WHERE text = \'1\'');
+        assert($query instanceof Statement);
```
[the remark this answers is not in the document]
Comment (jwage) [reply tests/Doctrine/Tests/DBAL/Functional/Schema/SchemaManagerFunctionalTestCase.php]: Change the `Doctrine\DBAL\Connection::query()` return type?

File: lib/Doctrine/DBAL/Query/Expression/ExpressionBuilder.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (morozov) [reply lib/Doctrine/DBAL/Query/Expression/ExpressionBuilder.php]: Wherever an argument represents an SQL expression, it should be a `string`.

File: tests/Doctrine/Tests/DBAL/Driver/StatementIteratorTest.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (morozov) [reply tests/Doctrine/Tests/DBAL/Driver/StatementIteratorTest.php]: If you change the expected argument type in `assertIterationCallsFetchOncePerStep()` from `Traversable` to `iterable`, it should work without additional changes.

File: tests/Doctrine/Tests/DBAL/Functional/DataAccessTest.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (morozov) [reply tests/Doctrine/Tests/DBAL/Functional/DataAccessTest.php]: `bindValue()` of a driver statement supports only `int` type but the DBAL statement which implements driver statement (which is a big question) supports also supports at least `string` and `Type` as `$type`. PHPStan should be given proper info on the statement type (dbal statement).

As for the value, all implementations accept `mixed $value`, so it must be some mistake.

File: tests/Doctrine/Tests/DBAL/Functional/Driver/IBMDB2/DB2StatementTest.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (morozov) [reply tests/Doctrine/Tests/DBAL/Functional/Driver/IBMDB2/DB2StatementTest.php]: I updated the link but it's not that important. Yes, given the number of assertions like this in the patch, we should make it clear that a DBAL Connection returns a DBAL Statement.

File: tests/Doctrine/Tests/DBAL/Functional/Schema/SchemaManagerFunctionalTestCase.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (morozov) [reply tests/Doctrine/Tests/DBAL/Functional/Schema/SchemaManagerFunctionalTestCase.php]: If possible, yes. Wherever a DBAL Connection returns a DBAL Statement, it should be clear from the API because regardless of the fact that the Driver and DBAL layer share the same interfaces (which is bad), they are not equal (that's why it's bad).

File: tests/Doctrine/Tests/DBAL/Driver/StatementIteratorTest.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (jwage) [reply tests/Doctrine/Tests/DBAL/Driver/StatementIteratorTest.php]: I can't get it to work. I fix phpcs, then phpstan complains and I can't get a type hint to work for all the different cases where we pass a mock object through. I'll fix one case and then it breaks it for the other.

```
FILE: tests/Doctrine/Tests/DBAL/Driver/StatementIteratorTest.php
-----------------------------------------------------------------------------------------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
-----------------------------------------------------------------------------------------------------------------------------------------------------
 102 | ERROR | Method \Doctrine\Tests\DBAL\Driver\StatementIteratorTest::assertIterationCallsFetchOncePerStep() does not have @param annotation for
     |       | its traversable parameter $iterator.
     |       | (SlevomatCodingStandard.TypeHints.TypeHintDeclaration.MissingTraversableParameterTypeHintSpecification)
-----------------------------------------------------------------------------------------------------------------------------------------------------
```

Then I add the type:

```php
    /**
     * @param iterable<int, mixed> $iterator
     */
    private function assertIterationCallsFetchOncePerStep(iterable $iterator, int &$calls) : void
    {
        foreach ($iterator as $i => $_) {
            $this->assertEquals($i + 1, $calls);
        }
    }
```

Then I get:

```
 ------ -----------------------------------------------------------------------------------------------------------------------------------
  Line   tests/Doctrine/Tests/DBAL/Driver/StatementIteratorTest.php
 ------ -----------------------------------------------------------------------------------------------------------------------------------
  61     Parameter #1 $iterator of method Doctrine\Tests\DBAL\Driver\StatementIteratorTest::assertIterationCallsFetchOncePerStep() expects
         iterable<int, mixed>, iterable&PHPUnit\Framework\MockObject\MockObject given.
 ------ -----------------------------------------------------------------------------------------------------------------------------------
```

File: tests/Doctrine/Tests/DBAL/Functional/Driver/IBMDB2/DB2StatementTest.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (jwage) [reply tests/Doctrine/Tests/DBAL/Functional/Driver/IBMDB2/DB2StatementTest.php]: This seems like it is out of the scope of this PR to make BC breaks in the core API of the DBAL.

File: tests/Doctrine/Tests/DBAL/Functional/DataAccessTest.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (jwage) [reply tests/Doctrine/Tests/DBAL/Functional/DataAccessTest.php]: This is the error when I try to bind a DateTime:

```
1) Doctrine\Tests\DBAL\Functional\DataAccessTest::testPrepareQueryBindValueDateTimeType
Object of class DateTime could not be converted to string

/data/repositories/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php:75
/data/repositories/dbal/lib/Doctrine/DBAL/Statement.php:113
/data/repositories/dbal/tests/Doctrine/Tests/DBAL/Functional/DataAccessTest.php:480
```

What is supposed to be responsible for converting the DateTime?

File: tests/Doctrine/Tests/DBAL/Functional/Driver/IBMDB2/DB2StatementTest.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (morozov) [reply tests/Doctrine/Tests/DBAL/Functional/Driver/IBMDB2/DB2StatementTest.php]: It doesn't have to be a breaking change. What I had in mind is adding a `@return` annotation which would hopefully override the type hint from the SA perspective but won't be enforced by the compiler.

File: tests/Doctrine/Tests/DBAL/Functional/DataAccessTest.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (morozov) [reply tests/Doctrine/Tests/DBAL/Functional/DataAccessTest.php]: It should be here and it seems to have worked before these changes:

https://github.com/doctrine/dbal/blob/d07906fddf156df4a3de8df28764e1485e4c885c/lib/Doctrine/DBAL/Statement.php#L102-L111

File: tests/Doctrine/Tests/DBAL/Functional/Driver/IBMDB2/DB2StatementTest.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (jwage) [reply tests/Doctrine/Tests/DBAL/Functional/Driver/IBMDB2/DB2StatementTest.php]: That still feels out of the scope of enabling phpstan in the tests folder. I don't mind making that change but I want to do it in an isolated PR without all these other changes muddying the review and testing of the change.

File: tests/Doctrine/Tests/DBAL/Functional/DataAccessTest.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (jwage) [reply tests/Doctrine/Tests/DBAL/Functional/DataAccessTest.php]: The API does not allow this to happen. This used to be possible:

```php
$stmt->bindValue(1, new DateTime('2010-01-01 10:10:10'), Types::DATETIME_MUTABLE);
```

Now that we enabled phpstan in the tests, we get this error:

```
 ------ -----------------------------------------------------------------------------------------------------
  Line   tests/Doctrine/Tests/DBAL/Functional/DataAccessTest.php
 ------ -----------------------------------------------------------------------------------------------------
  480    Parameter #3 $type of method Doctrine\DBAL\Driver\Statement::bindValue() expects int, string given.
 ------ -----------------------------------------------------------------------------------------------------
```

Is the API wrong, or do we need to always pass an `int` constant from `ParameterType` now?

[the remark this answers is not in the document]
Comment (morozov) [reply tests/Doctrine/Tests/DBAL/Functional/DataAccessTest.php]: It's the same problem as with the previous assertion. PHPStan thinks you have a driver statement which only accepts `int` as a parameter while you're dealing with a DBAL statement which accepts other types too.

[the remark this answers is not in the document]
Comment (jwage) [reply tests/Doctrine/Tests/DBAL/Functional/DataAccessTest.php]: Ok, I understand now. I see why you want to make the API change. I will try that.

File: tests/Doctrine/Tests/DBAL/Functional/Driver/IBMDB2/DB2StatementTest.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (morozov) [reply tests/Doctrine/Tests/DBAL/Functional/Driver/IBMDB2/DB2StatementTest.php]: I see your point. It's okay as long as it doesn't make you rework the tests which do work and cover valid cases like the one with `DateTime`. If you can fix those tests without changing their logic, then assertions like this can stay.

File: tests/Doctrine/Tests/DBAL/Functional/DataAccessTest.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (jwage) [reply tests/Doctrine/Tests/DBAL/Functional/DataAccessTest.php]: Hmm, I might need your help to look at this closer.  When I make the change you suggest I get this error:

```
 ------ ----------------------------------------------------------------------------------------------------------
  Line   lib/Doctrine/DBAL/Portability/Connection.php
 ------ ----------------------------------------------------------------------------------------------------------
  114    Method Doctrine\DBAL\Portability\Connection::prepare() should return Doctrine\DBAL\Statement but returns
         Doctrine\DBAL\Portability\Statement.
 ------ ----------------------------------------------------------------------------------------------------------
```

It opens up a can of worms I am not educated enough on to make decisions on what to do. The portability stuff is new since I last worked on DBAL.

[the remark this answers is not in the document]
Comment (morozov) [reply tests/Doctrine/Tests/DBAL/Functional/DataAccessTest.php]: It will require the segregation of the DBAL interfaces from the driver ones. I had this in mind for a long time and this is one more reason to do that.

Please suppress this failure in the PHPStan config instead of fixing it. The API is wrong.

[the remark this answers is not in the document]
Comment (jwage) [reply tests/Doctrine/Tests/DBAL/Functional/DataAccessTest.php]: Done.

[the remark this answers is not in the document]
Comment (morozov) [reply tests/Doctrine/Tests/DBAL/Functional/DataAccessTest.php]: I've filed #3591.

File: tests/Doctrine/Tests/DBAL/Functional/Schema/SchemaManagerFunctionalTestCase.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (morozov) [reply tests/Doctrine/Tests/DBAL/Functional/Schema/SchemaManagerFunctionalTestCase.php]: Cannot be done due to https://github.com/doctrine/dbal/issues/3591.

**Comment by morozov** [comment]
I only noticed now that the PR targeted against `develop`. Since it's not supposed to contain breaking changes, should we retarget it against `master`?

**Comment by jwage** [comment]
I can try and see how hard it will be to rebase against master.

File: tests/Doctrine/Tests/DBAL/Driver/StatementIteratorTest.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (morozov) [reply tests/Doctrine/Tests/DBAL/Driver/StatementIteratorTest.php]: This it what worked for me:
```diff
$ git diff upstream/develop...pull/3582 tests/Doctrine/Tests/DBAL/Driver/StatementIteratorTest.php
diff --git a/tests/Doctrine/Tests/DBAL/Driver/StatementIteratorTest.php b/tests/Doctrine/Tests/DBAL/Driver/StatementIteratorTest.php
index 63cc84c41..08787527c 100644
--- a/tests/Doctrine/Tests/DBAL/Driver/StatementIteratorTest.php
+++ b/tests/Doctrine/Tests/DBAL/Driver/StatementIteratorTest.php
@@ -15,7 +15,6 @@ use Doctrine\DBAL\Portability\Statement as PortabilityStatement;
 use Doctrine\Tests\DbalTestCase;
 use IteratorAggregate;
 use PHPUnit\Framework\MockObject\MockObject;
-use Traversable;
 use function extension_loaded;
 
 class StatementIteratorTest extends DbalTestCase
@@ -51,6 +50,7 @@ class StatementIteratorTest extends DbalTestCase
      */
     public function testStatementIterationCallsFetchOncePerStep(string $class) : void
     {
+        /** @var iterable<int, mixed>|MockObject $stmt */
         $stmt = $this->createPartialMock($class, ['fetch']);
 
         $calls = 0;
@@ -73,7 +73,10 @@ class StatementIteratorTest extends DbalTestCase
             });
     }
 
-    private function assertIterationCallsFetchOncePerStep(Traversable $iterator, int &$calls) : void
+    /**
+     * @param iterable<int, mixed> $iterator
+     */
+    private function assertIterationCallsFetchOncePerStep(iterable $iterator, int &$calls) : void
     {
         foreach ($iterator as $i => $_) {
             $this->assertEquals($i + 1, $calls);
```

**Comment by jwage** [comment]
I don't think doing this against master will work because the changes made to the tests depend on all the type and api work that we did in develop.

**Comment by morozov** [comment]
At least we tried. Maybe it could be done later. It still would make sense to get it in `master` because this is where we get most of the code changes before `3.0` is released.

**Review by morozov — APPROVED** [review]
